✨ Raise `AASM::UndefinedState` when calling `fire` or `fire!` with undefined event names

danielvdao opened the pull request:
When firing events that are unknown, we should check the state machine if the events exist before doing so, otherwise we'll get a `NoMethodError`.

Note - I saw the original PR here: https://github.com/aasm/aasm/pull/639 and wanted to see if I could play around and get it to pass 🙂 Also I recently started using the gem, so I was looking for a way to learn the gem better by contributing. Let me know if this isn't correct and I'll be happy to close it 🙇🏽‍♂️

danielvdao commented:
@anilmaurya can you trigger CI build since you were the original PR reviewer? I'm guessing opening a PR doesn't kick off the build like I hoped. I ran all the unit tests locally and it looks like this passed 🙏🏽

danielvdao commented:
Nudge @anilmaurya. Do you mind taking a look at this?

anilmaurya commented:
@danielvdao Sorry for late response.

I have triggered CI build, Build is passing.

I will review this PR in couple of days.

danielvdao commented:
@anilmaurya sounds good, hopefully should be a small change. Thank you!

anilmaurya commented on lib/aasm/instance_base.rb:
```diff
@@ -92,6 +91,7 @@ def permitted_transitions
     def state_object_for_name(name)
       obj = @instance.class.aasm(@name).states.find {|s| s.name == name}
       raise AASM::UndefinedState, "State :#{name} doesn't exist" if obj.nil?
+
```
I think this line addition is not required

anilmaurya commented on lib/aasm/instance_base.rb:
```diff
@@ -115,19 +115,25 @@ def may_fire_event?(name, *args)
-      event_name = event_name.to_s.+("!").to_sym
-      @instance.send(event_name, *args, &block)
+      bang_event_name = "#{event_name}!".to_sym
+      event = @instance.class.aasm(@name).state_machine.events[event_name]
+
+      raise AASM::UndefinedState, "State :#{bang_event_name} doesn't exist" if event.nil?
```
I think we can avoid duplicate code from `fire` method

danielvdao replied:
@anilmaurya sure. Are you suggesting I extract this into another method then?

anilmaurya commented on spec/unit/complex_example_spec.rb:
```diff
@@ -90,4 +90,12 @@
+
+  it "should raise AASM::UndefinedState when firing unknown bang events" do
+    expect { auth.aasm.fire(:unknown!) }.to raise_error(AASM::UndefinedState, "State :unknown! doesn't exist")
+  end
```
We should add specs to make sure that callbacks are not triggered when `UndefinedState` exception is raised

danielvdao replied:
Sounds good - I'll see if I can understand how callbacks get triggered, it will be fun 🙂

anilmaurya reviewed (changes requested):
Hi @danielvdao 

I added some review comments, let me know if you have any questions

danielvdao commented:
@anilmaurya I cleaned up a bit of lines and removed some debugger statements. I also added in more tests like you suggested. Let me know what you think.
